File: src/lib/studentSchedule.ts
import {
  EXAM_BLOCK_TYPES, MODULE_BLOCK_LABEL,
  type ModuleScheduleBlock, type ModuleScheduleStore,
} from '../data/moduleSchedule.ts'
import { moduleKey } from '../data/moduleSubjects.ts'
import type { UniYear, University } from '../data/universities.ts'

/**
 * A block on the student's timetable, with the module it belongs to attached.
 *
 * The admin console stores schedules per module, keyed
 * `<universityId>:<yearId>:<courseId>`. A student does not think in modules
 * when they look at a Tuesday, so this flattens every module in their year into
 * one list and keeps the module name for the label.
 */
export interface ScheduledSession extends ModuleScheduleBlock {
  courseId: string
  courseName: string
  /** Local `Date` for the block's start, or its date at midnight when untimed. */
  start: Date
  /** Null for a logbook task, which is a deadline rather than a sitting. */
  end: Date | null
  label: string
  isExam: boolean
}

/** `YYYY-MM-DD` and `HH:MM` as a local Date — never as UTC, which shifts the day. */
function localDateTime(date: string, time: string): Date | null {
  const [year, month, day] = date.split('-').map(Number)
  if (!year || !month || !day) return null
  const [hour, minute] = (time || '00:00').split(':').map(Number)
  return new Date(year, month - 1, day, hour || 0, minute || 0)
}

/**
 * The blocks the console published for one module of this student's year.
 *
 * Keyed by the year's **ID**, through the very `moduleKey` the console writes
 * with, so a reader and a writer of the same store can never drift apart again.
 * They had drifted: both once keyed by the year's *label*, and when years grew
 * stable IDs the console moved to `moduleKey(uni.id, y.id, c.id)` while this
 * reader was left building `<uni>:<year label>:<course>`. "KAU_Y1" and "Year 1"
 * never match, so every published timetable came back empty — no lectures and
 * no exams in the calendar, `nextExam` null, and therefore no exam horizon for
 * the adaptive planner, which quietly gave every student the no-exam shares.
 *
 * The label form is still read when the ID finds nothing, because schedules
 * published before that migration are genuinely stored under it. Preferring the
 * ID means current data always wins; iOS reads both for the same reason
 * (`ios/Synapse/Core/Model/StudentSchedule.swift`).
 */
function blocksFor(
  schedules: ModuleScheduleStore, university: University, year: UniYear, courseId: string,
): ModuleScheduleBlock[] {
  return schedules[moduleKey(university.id, year.id, courseId)]
    ?? schedules[moduleKey(university.id, year.year, courseId)]
    ?? []
}

/**
 * The student's own timetable, as their university published it.
 *
 * Empty until an admin has set up their year — which is the honest state, and
 * the reason the old generated calendar had to go: it manufactured lectures,
 * seminars and a recurring exam for every month, forever, for everyone.
 */
export function flattenSchedule(
  university: University, year: UniYear, schedules: ModuleScheduleStore,
): ScheduledSession[] {
  const sessions: ScheduledSession[] = []
  for (const course of year.courses) {
    for (const block of blocksFor(schedules, university, year, course.id)) {
      const start = localDateTime(block.date, block.startTime)
      if (!start) continue
      sessions.push({
        ...block,
        courseId: course.id,
        courseName: course.name,
        start,
        end: block.endTime ? localDateTime(block.date, block.endTime) : null,
        label: MODULE_BLOCK_LABEL[block.type] ?? block.type,
        isExam: EXAM_BLOCK_TYPES.includes(block.type),
      })
    }
  }
  sessions.sort((a, b) => a.start.getTime() - b.start.getTime())
  return sessions
}

/** Whether two dates fall on the same local calendar day. */
export function sameDay(a: Date, b: Date): boolean {
  return a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth() && a.getDate() === b.getDate()
}

/** Minutes a block occupies. Zero for an untimed logbook task. */
export function sessionMinutes(session: ScheduledSession): number {
  if (!session.end) return 0
  return Math.max(0, Math.round((session.end.getTime() - session.start.getTime()) / 60_000))
}

/**
 * The next exam on the timetable, and how many days away it is.
 *
 * Null when the student's year has no exam scheduled. The dashboard used to
 * claim "38 days to your Cardiovascular exam" from a literal; if no exam has
 * been published, the honest answer is to say nothing at all.
 */
export function nextExam(sessions: ScheduledSession[], now = new Date()): { session: ScheduledSession; daysAway: number } | null {
  const upcoming = sessions.find((session) => session.isExam && session.start.getTime() >= now.getTime())
  if (!upcoming) return null
  const midnightToday = new Date(now.getFullYear(), now.getMonth(), now.getDate())
  const midnightExam = new Date(upcoming.start.getFullYear(), upcoming.start.getMonth(), upcoming.start.getDate())
  return {
    session: upcoming,
    daysAway: Math.round((midnightExam.getTime() - midnightToday.getTime()) / 86_400_000),
  }
}

File: ios/Nishany/Core/Model/StudentSchedule.swift
import Foundation

/// The student's own timetable, as their university published it.
///
/// A port of `src/lib/useStudentSchedule.ts`. The admin console stores
/// schedules per module, keyed `<universityId>:<yearId>:<courseId>`. A student
/// does not think in modules when they look at a Tuesday, so this flattens
/// every module in their year into one list.
///
/// The key is looked up under the year's ID *and* its label. The console writes
/// `moduleKey(uni.id, y.id, c.id)` — the ID, e.g. `kau:KAU_Y1:…` — while the
/// website's own student reader builds the key from the year's label, so on the
/// web a published timetable never reaches the student it was published for.
/// Accepting both is not indecision: it is the only way to read the data that
/// actually exists while staying correct if that reader is ever fixed.
///
/// Empty until an admin has set up their year, which is the honest state: the
/// web deleted a generated calendar for exactly this reason, because it
/// manufactured lectures, seminars and a recurring exam for everyone, forever.
struct ScheduledSession: Identifiable, Equatable, Sendable {
    var id: String
    var type: String
    var title: String
    /// `YYYY-MM-DD`.
    var date: String
    var startTime: String
    var endTime: String
    var location: String
    var courseId: String
    var courseName: String
    /// Local, never UTC — reading a timetable as UTC shifts the day.
    var start: Date
    /// Nil for a logbook task, which is a deadline rather than a sitting.
    var end: Date?
    var label: String
    var isExam: Bool
}

enum StudentSchedule {

    /// The block types that are actually an exam.
    static let examTypes: Set<String> = ["midterm", "midyear", "term", "final"]

    static let label: [String: String] = [
        "lecture": "Lecture",
        "practical": "Practical session",
        "review": "Review session",
        "midterm": "Mid-term exam",
        "midyear": "Mid-year exam",
        "term": "Term exam",
        "final": "Final exam",
        "logbook": "Logbook task",
    ]

    /// Flatten the published schedules for this student's year.
    ///
    /// Both documents are read as loose JSON rather than decoded into types:
    /// they are admin-authored and grow fields regularly, and a timetable that
    /// refuses to load because a new key appeared is worse than one that
    /// ignores what it does not recognise.
    static func sessions(
        universities: Any?, schedules: Any?, audience: StudentAudience, now: Date = Date()
    ) -> [ScheduledSession] {
        guard audience.isKnown,
              let universities = universities as? [[String: Any]],
              let schedules = schedules as? [String: Any],
              let university = universities.first(where: { $0["id"] as? String == audience.universityId }),
              let years = university["years"] as? [[String: Any]],
              let year = years.first(where: { ($0["year"] as? String) == audience.year })
        else { return [] }

        let universityId = university["id"] as? String ?? audience.universityId
        let yearKeys = [year["id"] as? String, year["year"] as? String ?? audience.year]
            .compactMap { $0 }
        var out: [ScheduledSession] = []

        for course in year["courses"] as? [[String: Any]] ?? [] {
            guard let courseId = course["id"] as? String else { continue }
            let blocks = yearKeys
                .lazy
                .compactMap { schedules["\(universityId):\($0):\(courseId)"] as? [[String: Any]] }
                .first ?? []

            for block in blocks {
                guard let date = block["date"] as? String else { continue }
                let type = block["type"] as? String ?? "lecture"
                let startTime = block["startTime"] as? String ?? ""
                guard let start = localDateTime(date: date, time: startTime) else { continue }
                let endTime = block["endTime"] as? String ?? ""

                out.append(ScheduledSession(
                    id: block["id"] as? String ?? "\(courseId)-\(date)-\(startTime)",
                    type: type,
                    title: block["title"] as? String ?? "",
                    date: date,
                    startTime: startTime,
                    endTime: endTime,
                    location: block["location"] as? String ?? "",
                    courseId: courseId,
                    courseName: course["name"] as? String ?? courseId,
                    start: start,
                    end: endTime.isEmpty ? nil : localDateTime(date: date, time: endTime),
                    label: label[type] ?? type,
                    isExam: examTypes.contains(type)
                ))
            }
        }

        return out.sorted { $0.start < $1.start }
    }

    /// The next exam on the timetable, and how many days away it is.
    ///
    /// Nil when the student's year has no exam scheduled. The dashboard used to
    /// claim "38 days to your Cardiovascular exam" from a literal; if no exam
    /// has been published, the honest answer is to say nothing at all.
    static func nextExam(_ sessions: [ScheduledSession], now: Date = Date())
        -> (session: ScheduledSession, daysAway: Int)? {
        guard let upcoming = sessions.first(where: { $0.isExam && $0.start >= now }) else { return nil }

        // Measured between midnights, so an exam at nine tomorrow morning is
        // one day away rather than zero.
        let calendar = Calendar.current
        let today = calendar.startOfDay(for: now)
        let examDay = calendar.startOfDay(for: upcoming.start)
        let days = calendar.dateComponents([.day], from: today, to: examDay).day ?? 0
        return (upcoming, days)
    }

    /// `YYYY-MM-DD` and `HH:MM` as a local date — never as UTC, which shifts
    /// the day for anyone east or west of Greenwich.
    ///
    /// Shared with the upcoming list, which reads the student's own calendar
    /// and has to place its blocks on the same day this places sessions.
    static func localDateTime(date: String, time: String) -> Date? {
        let day = date.split(separator: "-").compactMap { Int($0) }
        guard day.count == 3 else { return nil }
        let clock = (time.isEmpty ? "00:00" : time).split(separator: ":").compactMap { Int($0) }

        var components = DateComponents()
        components.year = day[0]
        components.month = day[1]
        components.day = day[2]
        components.hour = clock.first ?? 0
        components.minute = clock.count > 1 ? clock[1] : 0
        return Calendar.current.date(from: components)
    }
}
